web3/eth_api_types.nim

```diff
@@ -299,6 +300,37 @@ func payload*(args: TransactionArgs): seq[byte] =
 func isEIP4844*(args: TransactionArgs): bool =
   args.maxFeePerBlobGas.isSome or args.blobVersionedHashes.isSome
 
+macro makeEthereumTypeEnum(): untyped =
+  ## This macro creates all the various types of Solidity contracts and maps
+  ## them to the type used for their encoding. It also creates an enum to
+  ## identify these types in the contract signatures, along with encoder
+  ## functions used in the generated procedures.
+  result = newStmtList()
+  var lastpow2: int
+  for i in countdown(256, 8, 8):
+    let
+      identUint = newIdentNode("EthereumUint" & $i)
+      identInt = newIdentNode("EthereumInt" & $i)
+    if ceil(log2(i.float)) == floor(log2(i.float)):
+      lastpow2 = i
+
+    result.add quote do:
+      type
+        `identUint`* = StUint[`lastpow2`]
```

Comment:
This in general will create some compatible-looking types in Nim which aren't actually compatible Solidity types, e.g., `int40` and `int48` are not the same in Solidity but they would end up being effectively the same in this mapping. Is that intention? Does it matter?

But this wouldn't hold between `int64` and `int72`: those would map to 64 and 128 respectively, and would not be the same type. So this also becomes a question of, is that ok for your use case? It creates sort of distinctions in the Nim-side integer types which are different than the Solidity ones.

Reply:
Also, this always creates `StInt`/`StUint` types, even for `lastpow2 <= 64`. Maybe not optimal.

Reply:
> would end up being effectively the same in this mapping

This might be an actual problem - ie we should give a compile-time error rather than being incompatible at runtime for "unsupported" integers (or support them properly)